google-cloud-clients/google-cloud-firestore/src/main/java/com/google/cloud/firestore/CollectionReference.java

```diff
@@ -167,6 +167,27 @@ public void remove() {
     };
   }
 
+  /**
+   * Retrieves the list of documents in this collection
+   *
+   * @return An ApiFuture
+   */
+  @Nonnull
+  public ApiFuture<?> listDocumentAsync() {
+    ListDocumentsRequest.Builder request = ListDocumentsRequest.newBuilder();
+    request.setParent(options.getParentPath().toString());
+    request.setCollectionId(options.getCollectionId());
+    request.setMask(DocumentMask.getDefaultInstance());
+    request.setShowMissing(true);
+
+    try {
+      return firestore.sendRequest(
+          request.build(), firestore.getClient().listDocumentsPagedCallable());
+    } catch (ApiException exception) {
+      throw FirestoreException.apiException(exception);
+    }
+  }
+
```

Comment:
The original API is already somewhat async. It returns an Iterator right away, but it does block on the "hasNext/next" call.  If you want to move away from the Iterator, then we have to expose `UnaryCallable<ListCollectionIdsRequest, ListCollectionIdsResponse>` (the type that the question mark refers to) and you can no longer use `iterateAll()`.

I understand that there are some usages for which this high-level API is not the right fit. There is however a middle-ground between using this SDK and the RPC layer - have you looked into using the auto-generated client that is exposed by https://github.com/googleapis/google-cloud-java/blob/10070bd8c646780363da6c21fe3d0f358be0b724/google-cloud-clients/google-cloud-firestore/src/main/java/com/google/cloud/firestore/v1/FirestoreClient.java

Reply:
@schmidt-sebastian yes right but i have question regarding `UnaryCallable<ListCollectionIdsRequest,ListCollectionIdsResponse>` if we expose `UnaryCallable<ListCollectionIdsRequest,ListCollectionIdsResponse>` then it will only returns collectionIdsList  there is no way to get list of document so how we can move ahead with this callable if we want to get list of document?

Reply:
You have to run it in two stages and use ListDocuments in the second stage:

https://github.com/googleapis/google-cloud-java/blob/10070bd8c646780363da6c21fe3d0f358be0b724/google-cloud-clients/google-cloud-firestore/src/main/java/com/google/cloud/firestore/v1/FirestoreClient.java#L262

If you want to iterate a Firestore database you have to switch recursively between these two calls.

Reply:
@schmidt-sebastian could please verify below code snippest and please suggest me if any changes required.
`
public ApiFuture<ListDocumentsPagedResponse> listDocumentAsync() throws  Exception {

      ListCollectionIdsRequest.Builder collectionIdsRequest = ListCollectionIdsRequest.newBuilder().setParent(options.getParentPath().toString());
      ListCollectionIdsResponse response = firestore.getClient().listCollectionIdsCallable().call(collectionIdsRequest.build());
      ApiFuture<ListDocumentsPagedResponse> listDocPagedResponseApiFuture = null;
      for(String id : response.getCollectionIdsList()){
          ListDocumentsRequest docRequest =
                  ListDocumentsRequest.newBuilder()
                          .setParent(options.getParentPath().toString())
                          .setCollectionId(id).build();
          listDocPagedResponseApiFuture= firestore.sendRequest(docRequest, firestore.getClient().listDocumentsPagedCallable());
      }
      return listDocPagedResponseApiFuture;
  }`